### ISSUE TITLE
editorconfig-set-coding-system runs mode hooks multiple times

### ISSUE THREAD

**Issue by xendk** [editorconfig/editorconfig-emacs#242]
After #236 mode hooks is run multiple times, in my case three.  This causes lsp-mode to start three language servers.

Reverting to the commit before fixes the problem.

**Comment by 10sr** [comment]
I know there are cases where mode hooks run multiple times, but I did not notice that causes multiple language servers running! 
I agree this should be fixed...

Could you paste the content of .editorconfig?

**Comment by xendk** [comment]
```
# Drupal editor configuration normalization
# @see http://editorconfig.org/

# This is the top-most .editorconfig file; do not search in parent directories.
root = true

# All files.
[*]
end_of_line = LF
indent_style = space
indent_size = 2
charset = utf-8
trim_trailing_whitespace = true
insert_final_newline = true

[composer.{json,lock}]
indent_size = 4
```

Removing `charset` and `end_of_line` suppresses the tripple calling of hooks, which is obvious why when inspecting `editorconfig-set-coding-system`.

**Comment by tastytea** [comment]
Hi, not sure if this helps, but I have a recipe to reproduce the issue.

### How to reproduce

Install clangd, extract [this test repo](https://github.com/editorconfig/editorconfig-emacs/files/5942956/editorconfig-issue-242-testrepo.tar.gz) and start emacs with `emacs -q -l init-debug.el`

Open `main.cpp` (everything seems fine) and then `other.cpp` (multiple disconnects).

#### init-debug.el

``` elisp
;;; init-debug.el ---  -*- lexical-binding: t; -*-
;;; Commentary:
;;; Code:

;; straight init
(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))

;; Install use-package
(straight-use-package 'use-package)
(setq straight-use-package-by-default t)


(use-package lsp-mode
  :hook ((c++-mode . lsp)))

(use-package editorconfig
  :config (editorconfig-mode 1))

;;; init-debug.el ends here
```

#### \*Messages* buffer
```
Loading package...done
Loading /home/tastytea/.emacs.d/early-theme.el (source)...done
Local Ispell dictionary set to en_US
Starting new Ispell process /usr/bin/hunspell with en_US dictionary...done
For information about GNU Emacs and the GNU system, type C-h C-a.
Loading /home/tastytea/.emacs.d/recentf...done
Cleaning up the recentf list...done (0 removed)
  <Open main.cpp>
[yas] Prepared just-in-time loading of snippets successfully.
Local Ispell dictionary set to en_US
LSP :: Connected to [clangd:15814/starting].
LSP :: clangd:15814 initialized successfully
  <Open other.cpp>
Local Ispell dictionary set to en_US
LSP :: Connected to [clangd:15814].
LSP :: Disconnected
Showing all blocks ... done
Local Ispell dictionary set to en_US
LSP :: Connected to [clangd:15814].
LSP :: Disconnected
Showing all blocks ... done
Local Ispell dictionary set to en_US
LSP :: Connected to [clangd:15814]. [2 times]
```

**Comment by 10sr** [comment]
I started testing a new implementation of editorconfig-mode, in hope that the issue reported in #242, #244 and #246 would be solved.
This can be enabled by setting `(setq editorconfig--enable-20210221-testing t)` before enabling editorconfig-mode.

It is not so tested very well yet, but currently I'm using this impl several days without any problem. 
If you are ok with it, I'd be very glad if you try this new version and see if it works better.
Thanks!

**Comment by tastytea** [comment]
My issue is gone with `editorconfig--enable-20210221-testing`. Thank you!

**Comment by xendk** [comment]
Well, it's working so well I forgot about this issue. I've been running with it since and have had no problems.

**Comment by 10sr** [comment]
Yeah, Thanks! 👍 
However, I found there are a few corner-cases where it works poorly (especially when changing major-mode after visiting files), so I'm going to enable this by default after fixing this.
But recently I have no time to work on this...

**Comment by 10sr** [comment]
I just merged #263 so this is now enabled by default 🙌 (set editorconfig--legacy-version to disable this and use old version)
